src/main/kotlin/org/rust/lang/core/resolve2/DefMapsBuilder.kt

```diff
@@ -151,3 +160,43 @@ class DefMapsBuilder(
         }
     }
 }
+
+/**
+ * Needed because of [DefCollector.expandMacrosInParallel].
+ * We use [ForkJoinPool.invokeAll] there, which can execute tasks
+ * from completely different thread pool (associated with current thread).
+ * That's why we need to be sure that we build DefMaps on "fresh" thread.
+ * Also see [org.rust.lang.core.macros.MacroExpansionServiceImplInner.pool].
+ */
+private fun invokeWithoutHelpingOtherForkJoinPools(
+    fixedThreadPool: ExecutorService,
+    forkJoinPool: ExecutorService,
+    action: () -> Unit
+) {
+    /** Current thread is [ForkJoinWorkerThread]. */
+    check(fixedThreadPool !is ForkJoinPool)
+    fixedThreadPool.submit {
+        /** Now we are not in [ForkJoinWorkerThread]. */
+        val future = CompletableFuture<Unit>()
+        forkJoinPool.submit {
+            try {
+                action()
+                future.complete(Unit)
+            } catch (e: Throwable) {
+                future.completeExceptionally(e)
+            }
+        }
+        /**
+         * Waiting on [ForkJoinTask] will try to help either current thread or [ForkJoinPool.common].
+         *   ⇒ so we can't use `.submit { ... }.getWithRethrow()`
+         * Waiting on [CompletableFuture] will try to help only current thread.
+         *   ⇒ so we use it.
+         */
+        future.getWithRethrow()
+    }
+        /**
+         * This [getWithRethrow] will not help any [ForkJoinPool]
+         * because [fixedThreadPool] is usual [ThreadPoolExecutor].
+         */
+        .getWithRethrow()
```

Comment:
Maybe we could try to use `com.google.common.util.concurrent.SettableFuture` instead of `CompletableFuture`? It should just park the thread instead of helping `ForkJoinPool`.

We could even try to write another version `invokeAll` based on `SettableFuture` and remove all these hacks with thread delegation